### PR TITLE
AF-2933 : Business Cental fails to start with error "no marshaller for type: org.guvnor.structure.contributors.Contributor" after migrating to 7.12.0

### PR DESCRIPTION
**Pull request by akumar074**
- Fixing Contributor and ContributorType packages

**JIRA**: 

[AF-2933](https://issues.redhat.com/browse/AF-2933)
[RHPAM-4152](https://issues.redhat.com/browse/RHPAM-4152)

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

**Comment by akumar074** [comment]
Jenkins run fdb

**Comment by akumar074** [comment]
Jenkins run fdb

**Comment by sthundat** [comment]
@akumar074 
Hi Abhishek,

Thanks for the PR!

PR has fixed the Marshaling error and spaces/projects are successfully migrated to the new version(7.12.0) of BC using migration tool.

But I have observed below repository not found error with dashbuilder while starting business central with the migrated version.

`12:11:39,376 ERROR [org.uberfire.ext.metadata.io.BatchIndex] (Thread-114) Index fails. [@dashbuilder/datasets/master]: java.util.concurrent.CompletionException: java.lang.RuntimeException: Repository not found inside space dashbuilder with path default://master@dashbuilder/datasets/definitions/gitContributors.dset (root path default://master@dashbuilder/datasets/)
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:718)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1705)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)`
While migrating I have used one sample `pom-migration.json` file for pom migration with some repository details that I get it from the documentation.
Link for the documentation,
https://access.redhat.com/documentation/en-us/red_hat_process_automation_manager/7.4/html/patching_and_upgrading_red_hat_process_automation_manager_7.4/migration-tool-niogit-7.4-proc_patching-upgrading

2.Also when I try to create new project from BC ,that also leads to error
image

below is the eap log,
`12:15:49,965 ERROR [org.guvnor.structure.backend.repositories.RepositoryServiceImpl] (default task-5) Error while removing repositories: java.lang.RuntimeException: Repository not found
	at deployment.business-central.war//org.guvnor.structure.backend.organizationalunit.OrganizationalUnitServiceImpl.lambda$removeRepository$16(OrganizationalUnitServiceImpl.java:456)
	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
	at deployment.business-central.war//org.guvnor.structure.backend.organizationalunit.OrganizationalUnitServiceImpl.lambda$removeRepository$17(OrganizationalUnitServiceImpl.java:456)
	at deployment.business-central.war//org.guvnor.structure.backend.organizationalunit.config.SpaceConfigStorageRegistryImpl$SpaceStorageBatchImpl.run(SpaceConfigStorageRegistryImpl.java:104)
	at deployment.business-central.war//org.guvnor.structure.backend.organizationalunit.OrganizationalUnitServiceImpl.removeRepository(OrganizationalUnitServiceImpl.java:450)
	at deployment.business-central.war//org.guvnor.structure.backend.organizationalunit.OrganizationalUnitServiceImpl$Proxy$_$$_WeldClientProxy.removeRepository(Unknown Source)
	at deployment.business-central.war//org.guvnor.structure.backend.repositories.RepositoryServiceImpl.doRemoveRepository(RepositoryServiceImpl.java:460)
	at deployment.business-central.war//org.guvnor.structure.backend.repositories.RepositoryServiceImpl.lambda$removeRepositories$7(RepositoryServiceImpl.java:424)
	at deployment.business-central.war//org.guvnor.structure.backend.organizationalunit.config.SpaceConfigStorageRegistryImpl$SpaceStorageBatchImpl.run(SpaceConfigStorageRegistryImpl.java:104)
	at deployment.business-central.war//org.guvnor.structure.backend.repositories.RepositoryServiceImpl.removeRepositories(RepositoryServiceImpl.java:416)
	at deployment.business-central.war//org.guvnor.structure.backend.repositories.RepositoryServiceImpl$Proxy$_$$_WeldClientProxy.removeRepositories(Unknown Source)
	at deployment.business-central.war//org.guvnor.structure.backend.organizationalunit.OrganizationalUnitServiceImpl.removeOrganizationalUnit(OrganizationalUnitServiceImpl.java:523)
	at deployment.business-central.war//org.guvnor.structure.backend.organizationalunit.OrganizationalUnitServiceImpl$Proxy$_$$_WeldClientProxy.removeOrganizationalUnit(Unknown Source)
	at deployment.business-central.war//org.kie.workbench.common.screens.examples.backend.server.ExamplesServiceImpl.lambda$deleteOldHiddenSpaces$1(ExamplesServiceImpl.java:245)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at deployment.business-central.war//org.kie.workbench.common.screens.examples.backend.server.ExamplesServiceImpl.deleteOldHiddenSpaces(ExamplesServiceImpl.java:245)
	at deployment.business-central.war//org.kie.workbench.common.screens.examples.backend.server.ExamplesServiceImpl.lambda$initPlaygroundRepository$0(ExamplesServiceImpl.java:184)
	at deployment.business-central.war//org.guvnor.structure.backend.organizationalunit.config.SpaceConfigStorageRegistryImpl$SpaceStorageBatchImpl.run(SpaceConfigStorageRegistryImpl.java:104)
	at deployment.business-central.war//org.kie.workbench.common.screens.examples.backend.server.ExamplesServiceImpl.initPlaygroundRepository(ExamplesServiceImpl.java:179)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)`
3.Can you please add relevant tests for this issue?

Please let me know your inputs.

Thanks,
Savitha

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/organizationalunit/config/SpaceConfigStorageImpl.java [hunk 1]
```diff
@@ -292,7 +294,36 @@ BranchPermissions getDefaultBranchPermissions(String branchName) {
...
+                if (content.contains(OLD_CONTRIBUTOR_CLASS)) {
+                    objectStorage.write(path,
+                                        Marshalling.fromJSON(content.replaceAll(OLD_CONTRIBUTOR_CLASS,
+                                                                                NEW_CONTRIBUTOR_CLASS),
+                                                             SpaceInfo.class));
+                }
+                return objectStorage.read(path);
+            }
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
Comment (sthundat) [inline uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/organizationalunit/config/SpaceConfigStorageImpl.java:324]: I would prefer to pass message while creating the runtime exception

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/organizationalunit/config/SpaceConfigStorageImpl.java [hunk 2]
```diff
@@ -292,7 +294,36 @@ BranchPermissions getDefaultBranchPermissions(String branchName) {
...
     public SpaceInfo loadSpaceInfo() {
-        return objectStorage.read(buildSpaceConfigFilePath(SPACE_INFO));
+        try {
+            return objectStorage.read(buildSpaceConfigFilePath(SPACE_INFO));
+        } catch (RuntimeException e) {
+            if (e.getCause() instanceof MarshallingException) {
+                logger.error(e.getMessage());
+                return fixContributorClassInSpaceInfo();
+            }
+            throw new RuntimeException(e);
```
Comment (sthundat) [inline uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/organizationalunit/config/SpaceConfigStorageImpl.java:304]: I would prefer to pass  an exception message string while creating the runtime exception

**Comment by akumar074** [comment]
Jenkins run fdb

**Comment by sthundat** [comment]
@akumar074 
Hi Abhishek,
I observed that, the error I have mentioned above happening only when I first time starts the BC after migration.But it doesn't break any functionality of BC.Hence I would suggest  we can close this issue and I can create a separate Jira to track the above issue.
WDYT?
thanks,
Savitha

**Comment by akumar074** [comment]
Hi @sthundat 
I agree we can create a new Jira to track the above error. I'll add some tests for this and we can resolve this issue.
Thanks

**Comment by akumar074** [comment]
Jenkins run fdb

**Comment by akumar074** [comment]
Jenkins run fdb

**Comment by akumar074** [comment]
Jenkins run fdb

**Comment by akumar074** [comment]
Jenkins run cdb

**Comment by akumar074** [comment]
Jenkins run fdb

**Comment by akumar074** [comment]
Jenkins run fdb

**Comment by sthundat** [comment]
Jenkins run fdb

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/organizationalunit/config/SpaceConfigStorageImpl.java (hunk 2 above)
Comment (adrielparedes) [reply uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/organizationalunit/config/SpaceConfigStorageImpl.java:304]: +1, also we are not loggin the problem unless it is a MarshallingException problem.

**Comment by akumar074** [comment]
Jenkins run fdb

**Comment by akumar074** [comment]
Hi @sthundat @adrielparedes @RishiRajAnand I have made changes according to the comments. Please rereview the PR when you can.
Thanks

**Comment by sthundat** [comment]
Hi @akumar074 ,
Thanks for the changes.
FDB doesn't generate the migration tool.
Can you built migration tool for the final check
thanks,
Savitha

**Comment by akumar074** [comment]
@sthundat I will build the migration tool locally and share the zip with you

**Review by adrielparedes — APPROVED** [review]

**Comment by akumar074** [comment]
Jenkins run fdb

**Comment by akumar074** [comment]
Jenkins run cdb

**Comment by sthundat** [comment]
@akumar074 
Hi Abhishek,
Changes looks good to me.Created Jira for `repository not found exception`[Link](https://issues.redhat.com/browse/RHPAM-4205)
Thanks,
Savitha

**Review by RishiRajAnand — APPROVED** [review]

**Review by sthundat — APPROVED** [review]

**Comment by akumar074** [comment]
Hi @adrielparedes , We have all the approvals required with Green FDB build. Can we go ahead and merge this?
Thanks

**Comment by akumar074** [comment]
Jenkins run cdb

**Pull request merged**